File: sample-videochat-conference-java/src/main/java/com/quickblox/sample/videochat/conference/java/activities/SelectUsersActivity.java
package com.quickblox.sample.videochat.conference.java.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.SystemClock;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ListView;

import com.quickblox.chat.model.QBChatDialog;
import com.quickblox.core.QBEntityCallback;
import com.quickblox.core.exception.QBResponseException;
import com.quickblox.sample.videochat.conference.java.R;
import com.quickblox.sample.videochat.conference.java.adapters.CheckboxUsersAdapter;
import com.quickblox.sample.videochat.conference.java.db.QbUsersDbManager;
import com.quickblox.sample.videochat.conference.java.utils.Consts;
import com.quickblox.sample.videochat.conference.java.utils.SharedPrefsHelper;
import com.quickblox.sample.videochat.conference.java.utils.ToastUtils;
import com.quickblox.users.model.QBUser;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.TimeUnit;

import androidx.annotation.StringRes;
import androidx.fragment.app.Fragment;


public class SelectUsersActivity extends BaseActivity {
    public static final String EXTRA_QB_USERS = "qb_users";
    public static final String EXTRA_QB_OCCUPANTS_IDS = "qb_occupants_ids";
    public static final int MINIMUM_CHAT_OCCUPANTS_SIZE = 2;
    private static final long CLICK_DELAY = TimeUnit.SECONDS.toMillis(2);

    private static final String EXTRA_QB_DIALOG = "qb_dialog";

    private QbUsersDbManager dbManager;

    private ListView usersListView;
    private CheckboxUsersAdapter usersAdapter;
    private long lastClickTime = 0l;
    private QBUser currentUser;
    private QBChatDialog dialog;

    public static void startForResult(Fragment fragment, int code, QBChatDialog dialog) {
        Intent intent = new Intent(fragment.getContext(), SelectUsersActivity.class);

        intent.putExtra(EXTRA_QB_DIALOG, dialog);
        fragment.startActivityForResult(intent, code);
    }

    /**
     * Start activity for picking users
     *
     * @param activity activity to return result
     * @param code     request code for onActivityResult() method
     *                 <p>
     *                 in onActivityResult there will be 'ArrayList<QBUser>' in the intent extras
     *                 which can be obtained with SelectPeopleActivity.EXTRA_QB_USERS key
     */
    public static void startForResult(Activity activity, int code) {
        Intent intent = new Intent(activity, SelectUsersActivity.class);

        activity.startActivityForResult(intent, code);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_select_users);

        usersListView = findViewById(R.id.list_select_users);
        currentUser = sharedPrefsHelper.getQbUser();
        setActionbarTitle(isEditingChat() ? R.string.select_users_edit_dialog : R.string.select_users_create_dialog);

        actionBar.setDisplayHomeAsUpEnabled(true);

        dbManager = QbUsersDbManager.getInstance(getApplicationContext());
        initUsersAdapter();
    }

    @Override
    protected void onDestroy() {
        hideProgressDialog();
        super.onDestroy();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.activity_select_users, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if ((SystemClock.uptimeMillis() - lastClickTime) < CLICK_DELAY) {
            return super.onOptionsItemSelected(item);
        }
        lastClickTime = SystemClock.uptimeMillis();

        switch (item.getItemId()) {
            case R.id.menu_select_people_action_done:
                if (isEditingChat()) {
                    addOccupantsToDialog();
                } else if (usersAdapter != null) {
                    List<QBUser> users = usersAdapter.getSelectedUsers();
                    if (users.size() >= MINIMUM_CHAT_OCCUPANTS_SIZE) {
                        passResultToCallerActivity(null);
                    } else {
                        ToastUtils.shortToast(R.string.select_users_choose_users);
                    }
                }
                return true;
            case R.id.menu_refresh_users:
                if (isEditingChat()) {
                    updateDialogAndUsers();
                } else {
                    loadUsersFromQb();
                }
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void addOccupantsToDialog() {
        showProgressDialogIfPossible(R.string.dlg_updating_dialog);

        List<QBUser> users = usersAdapter.getSelectedUsers();
        QBUser[] usersArray = users.toArray(new QBUser[users.size()]);
        Log.d("SelectedUsersActivity", "usersArray= " + Arrays.toString(usersArray));

        requestExecutor.updateDialog(dialog, usersArray, new QBEntityCallback<QBChatDialog>() {
            @Override
            public void onSuccess(QBChatDialog dialog, Bundle params) {
                dismissProgressDialogIfPossible();
                passResultToCallerActivity(dialog.getOccupants());
            }

            @Override
            public void onError(QBResponseException responseException) {
                dismissProgressDialogIfPossible();
                showErrorSnackbar(R.string.dlg_updating_dialog, responseException, new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        addOccupantsToDialog();
                    }
                });
            }
        });
    }

    private void removeExistentOccupants(List<QBUser> users) {
        List<Integer> userIDs = dialog.getOccupants();
        if (userIDs == null) {
            return;
        }

        Iterator<QBUser> i = users.iterator();
        while (i.hasNext()) {
            QBUser user = i.next();

            for (Integer userID : userIDs) {
                if (user.getId().equals(userID)) {
                    Log.d("SelectedUsersActivity", "users.remove(user)= " + user);
                    i.remove();
                }
            }
        }
    }

    private void passResultToCallerActivity(List<Integer> occupantsIds) {
        Intent result = new Intent();
        ArrayList<QBUser> selectedUsers = new ArrayList<>(usersAdapter.getSelectedUsers());
        result.putExtra(EXTRA_QB_USERS, selectedUsers);
        if (occupantsIds != null) {
            result.putExtra(EXTRA_QB_OCCUPANTS_IDS, (Serializable) occupantsIds);
        }
        setResult(RESULT_OK, result);
        finish();
    }

    private void initUsersAdapter() {
        ArrayList<QBUser> users = dbManager.getAllUsers();
        dialog = (QBChatDialog) getIntent().getSerializableExtra(EXTRA_QB_DIALOG);
        if (dialog != null) {
            updateDialogAndUsers();
            usersAdapter = new CheckboxUsersAdapter(SelectUsersActivity.this, new ArrayList<QBUser>(), currentUser);
        } else {
            usersAdapter = new CheckboxUsersAdapter(SelectUsersActivity.this, users, currentUser);
        }
        usersListView.setAdapter(usersAdapter);
    }

    private boolean isEditingChat() {
        return getIntent().getSerializableExtra(EXTRA_QB_DIALOG) != null;
    }

    private void updateDialogAndUsers() {
        showProgressDialogIfPossible(R.string.dlg_loading_dialogs_users);
        requestExecutor.loadDialogByID(dialog.getDialogId(), new QBEntityCallback<QBChatDialog>() {
            @Override
            public void onSuccess(QBChatDialog dialog, Bundle params) {
                SelectUsersActivity.this.dialog.setOccupantsIds(dialog.getOccupants());
                loadUsersFromQb();
                dismissProgressDialogIfPossible();
            }

            @Override
            public void onError(QBResponseException responseException) {
                dismissProgressDialogIfPossible();
                showErrorSnackbar(R.string.loading_dialog_error, responseException, new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        loadUsersFromQb();
                    }
                });
            }
        });
    }

    private void showProgressDialogIfPossible(@StringRes int messageId) {
        if (!isFinishing()) {
            showProgressDialog(messageId);
        }
    }

    private void dismissProgressDialogIfPossible() {
        if (!isFinishing()) {
            hideProgressDialog();
        }
    }

    private void loadUsersFromQb() {
        showProgressDialogIfPossible(R.string.dlg_loading_dialogs_users);
        String currentRoomName = SharedPrefsHelper.getInstance().get(Consts.PREF_CURREN_ROOM_NAME);

        requestExecutor.loadUsersByTag(currentRoomName, new QBEntityCallback<ArrayList<QBUser>>() {
            @Override
            public void onSuccess(ArrayList<QBUser> users, Bundle params) {
                dbManager.saveAllUsers(users, true);
                if (isEditingChat()) {
                    users.remove(currentUser);
                    removeExistentOccupants(users);
                }
                usersAdapter.updateList(users);

                dismissProgressDialogIfPossible();
            }

            @Override
            public void onError(QBResponseException responseException) {
                dismissProgressDialogIfPossible();
                showErrorSnackbar(R.string.loading_users_error, responseException, new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        loadUsersFromQb();
                    }
                });
            }
        });
    }
}

File: sample-videochat-conference-kotlin/src/main/java/com/quickblox/sample/videochat/conference/kotlin/activities/SelectUsersActivity.kt
package com.quickblox.sample.videochat.conference.kotlin.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.SystemClock
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.quickblox.chat.model.QBChatDialog
import com.quickblox.core.QBEntityCallback
import com.quickblox.core.exception.QBResponseException
import com.quickblox.sample.videochat.conference.kotlin.R
import com.quickblox.sample.videochat.conference.kotlin.adapter.CheckboxUsersAdapter
import com.quickblox.sample.videochat.conference.kotlin.db.QbUsersDbManager
import com.quickblox.sample.videochat.conference.kotlin.util.loadDialogByID
import com.quickblox.sample.videochat.conference.kotlin.util.loadUsersByTag
import com.quickblox.sample.videochat.conference.kotlin.util.updateDialog
import com.quickblox.sample.videochat.conference.kotlin.utils.PREF_CURRENT_ROOM_NAME
import com.quickblox.sample.videochat.conference.kotlin.utils.SharedPrefsHelper
import com.quickblox.sample.videochat.conference.kotlin.utils.shortToast
import com.quickblox.users.model.QBUser
import java.io.Serializable
import java.util.*
import java.util.concurrent.TimeUnit


class SelectUsersActivity : BaseActivity() {

    private lateinit var currentUser: QBUser
    private lateinit var usersRecyclerView: RecyclerView

    private val clickDelay = TimeUnit.SECONDS.toMillis(2)
    private var lastClickTime = 0L
    private var dialog: QBChatDialog? = null

    private var usersAdapter: CheckboxUsersAdapter? = null

    companion object {
        private const val EXTRA_USERS_IDS_KEY = "users_ids_key"
        private const val EXTRA_DIALOG_KEY = "dialog_key"

        private const val MINIMUM_CHAT_OCCUPANTS_SIZE = 2

        const val RESULT_USERS_KEY = "result_users_key"

        fun startForResult(activity: Activity, code: Int) {
            activity.startActivityForResult(Intent(activity, SelectUsersActivity::class.java), code)
        }

        fun startForResult(fragment: Fragment, code: Int, dialog: QBChatDialog) {
            val intent = Intent(fragment.activity, SelectUsersActivity::class.java)
            intent.putExtra(EXTRA_DIALOG_KEY, dialog)
            fragment.startActivityForResult(intent, code)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_select_users)

        currentUser = SharedPrefsHelper.getQbUser()!!
        dialog = intent.getSerializableExtra(EXTRA_DIALOG_KEY) as QBChatDialog?

        initUI()
        initActionBar()
        initUsersAdapter()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }

        if (SystemClock.uptimeMillis() - lastClickTime < clickDelay) {
            return super.onOptionsItemSelected(item)
        }

        lastClickTime = SystemClock.uptimeMillis()

        when (item.itemId) {
            R.id.menu_select_people_action_done -> {
                if (dialog != null) {
                    addOccupantsToDialog(dialog!!)
                } else if (usersAdapter != null) {
                    val users = usersAdapter?.getSelectedUsers()!!
                    if (users.size >= MINIMUM_CHAT_OCCUPANTS_SIZE) {
                        passResultToCallerActivity(null)
                    } else {
                        shortToast(R.string.select_users_choose_users)
                    }
                }
                return true
            }
            R.id.menu_refresh_users -> {
                if (dialog != null) {
                    updateDialogAndUsers(dialog!!)
                } else {
                    loadUsersFromQb()
                }
                return super.onOptionsItemSelected(item)
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.activity_select_users, menu)
        return true
    }

    private fun initUI() {
        usersRecyclerView = findViewById(R.id.list_select_users)
    }

    private fun initActionBar() {
        val title = if (dialog != null) {
            getString(R.string.select_users_edit_dialog)
        } else {
            getString(R.string.select_users_create_dialog)
        }
        supportActionBar?.title = title
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
    }

    private fun initUsersAdapter() {
        val users = QbUsersDbManager.getAllUsers()
        if (dialog != null) {
            updateDialogAndUsers(dialog!!)
            usersAdapter = CheckboxUsersAdapter(this, ArrayList(), currentUser)
        } else {
            usersAdapter = CheckboxUsersAdapter(this, users, currentUser)
        }
        usersRecyclerView.layoutManager = LinearLayoutManager(this)
        usersRecyclerView.adapter = usersAdapter
    }

    private fun updateDialogAndUsers(dialog: QBChatDialog) {
        showProgressDialog(R.string.dlg_loading_dialogs_users)
        loadDialogByID(dialog.dialogId, object : QBEntityCallback<QBChatDialog> {
            override fun onSuccess(dialog: QBChatDialog, params: Bundle) {
                this@SelectUsersActivity.dialog?.setOccupantsIds(dialog.occupants)
                loadUsersFromQb()
                hideProgressDialog()
            }

            override fun onError(responseException: QBResponseException) {
                hideProgressDialog()
                showErrorSnackbar(R.string.dlg_error_loading, responseException, View.OnClickListener { loadUsersFromQb() })
            }
        })
    }

    private fun addOccupantsToDialog(dialog: QBChatDialog) {
        showProgressDialog(R.string.dlg_updating_dialog)

        val users = usersAdapter?.getSelectedUsers()
        val usersArray = users?.toTypedArray()
        Log.d("SelectedUsersActivity", "usersArray= " + Arrays.toString(usersArray))

        updateDialog(dialog, usersArray as Array<QBUser>, object : QBEntityCallback<QBChatDialog> {
            override fun onSuccess(dialog: QBChatDialog, params: Bundle) {
                hideProgressDialog()
                passResultToCallerActivity(dialog.occupants)
            }

            override fun onError(responseException: QBResponseException) {
                hideProgressDialog()
                showErrorSnackbar(R.string.dlg_updating_dialog, responseException, View.OnClickListener { addOccupantsToDialog(dialog) })
            }
        })
    }

    private fun passResultToCallerActivity(occupantsIds: List<Int>?) {
        val result = Intent()
        val selectedUsers = ArrayList(usersAdapter?.getSelectedUsers())
        result.putExtra(RESULT_USERS_KEY, selectedUsers)
        if (occupantsIds != null) {
            result.putExtra(EXTRA_USERS_IDS_KEY, occupantsIds as Serializable?)
        }
        setResult(Activity.RESULT_OK, result)
        finish()
    }

    private fun loadUsersFromQb() {
        showProgressDialog(R.string.dlg_loading_dialogs_users)
        val currentRoomName = SharedPrefsHelper.get<String>(PREF_CURRENT_ROOM_NAME)
        loadUsersByTag(currentRoomName, object : QBEntityCallback<ArrayList<QBUser>> {
            override fun onSuccess(users: ArrayList<QBUser>, params: Bundle) {
                QbUsersDbManager.saveAllUsers(users, true)
                if (dialog != null) {
                    users.remove(currentUser)
                    removeExistentOccupants(users, dialog!!)
                }
                usersAdapter?.updateUsers(users)
                hideProgressDialog()
            }

            override fun onError(responseException: QBResponseException) {
                hideProgressDialog()
                showErrorSnackbar(R.string.loading_users_error, responseException, View.OnClickListener { loadUsersFromQb() })
            }
        })
    }

    private fun removeExistentOccupants(users: MutableList<QBUser>, dialog: QBChatDialog) {
        val userIDs = dialog.occupants ?: return
        val i = users.iterator()
        while (i.hasNext()) {
            val user = i.next()
            for (userID in userIDs) {
                if (user.id == userID) {
                    Log.d("SelectedUsersActivity", "users.remove(user)= $user")
                    i.remove()
                }
            }
        }
    }
}